controllers/provisioners/eks/rollingupdate.go

```diff
@@ -0,0 +1,301 @@
+				needsUpdate = append(needsUpdate, instanceId)
+			}
+		}
+
+	}
+	allCount := len(allInstances)
+
+	var unavailableInt int
+	if maxUnavailable.Type == intstr.String {
+		unavailableInt, _ = intstr.GetValueFromIntOrPercent(maxUnavailable, allCount, true)
+	} else {
+		unavailableInt = maxUnavailable.IntValue()
+	}
+
+	if unavailableInt == 0 {
+		unavailableInt = 1
+	}
+
+	sort.Strings(needsUpdate)
+
+	return &RollingUpdateRequest{
+		Logger:           ctx.Log,
+		KubernetesClient: ctx.KubernetesClient,
+		AwsWorker:        ctx.AwsWorker,
+		DrainGroup:       ctx.DrainGroup,
+		ClusterNodes:     state.GetClusterNodes(),
+		MaxUnavailable:   unavailableInt,
+		DesiredCapacity:  desiredCount,
+		AllInstances:     allInstances,
+		UpdateTargets:    needsUpdate,
+		ScalingGroup:     scalingGroup,
+	}
+}
+
+func (r *RollingUpdateRequest) ProcessRollingUpgradeStrategy() (bool, error) {
+
+	var (
+		scalingGroupName = aws.StringValue(r.ScalingGroup.AutoScalingGroupName)
+	)
+
+	ok := awsprovider.IsDesiredInService(r.ScalingGroup)
+	if !ok {
+		r.Info("desired instances are not in service", "scalinggroup", scalingGroupName)
+		return false, nil
+	}
+
+	ok, err := kubeprovider.IsMinNodesReady(r.ClusterNodes, r.AllInstances, r.MaxUnavailable)
+	if err != nil {
+		return false, err
+	}
+
+	if !ok {
+		r.Info("desired nodes are not ready", "scalinggroup", scalingGroupName)
+		return false, nil
+	}
+
+	r.Info("starting rolling update", "scalinggroup", scalingGroupName, "targets", r.UpdateTargets, "maxunavailable", r.MaxUnavailable)
+
+	if len(r.UpdateTargets) == 0 {
+		r.Info("no updatable instances", "scalinggroup", scalingGroupName)
+		return true, nil
+	}
+
+	// cannot rotate if maxUnavailable is greater than number of desired
+	if r.MaxUnavailable > r.DesiredCapacity {
+		r.Info("maxUnavailable exceeds desired capacity, setting maxUnavailable match desired",
+			"scalinggroup", scalingGroupName, "maxunavailable", r.MaxUnavailable, "desiredcapacity", r.DesiredCapacity)
+		r.MaxUnavailable = r.DesiredCapacity
+	}
+
+	var terminateTargets []string
+	if r.MaxUnavailable <= len(r.UpdateTargets) {
+		terminateTargets = r.UpdateTargets[:r.MaxUnavailable]
+	} else {
+		terminateTargets = r.UpdateTargets
+	}
+
+	targetNodes := kubeprovider.GetNodesByInstance(terminateTargets, r.ClusterNodes)
+
+	targetNames := make([]string, 0)
+	for _, node := range targetNodes.Items {
+		targetNames = append(targetNames, node.GetName())
+	}
+
+	r.Info("starting rotation on target nodes", "targets", targetNames)
+
+	// Only create new threads if waitgroup is empty
+	drainErrs := make(chan error)
+	if reflect.DeepEqual(r.DrainGroup, &sync.WaitGroup{}) {
+		for _, node := range targetNodes.Items {
+
+			nodeName := node.GetName()
+			r.Info("creating drainer goroutine for node", "node", nodeName)
+
+			drainOpts := &drain.Helper{
+				DeleteLocalData:     true,
+				Force:               true,
```

Comment:
Might consider making force (and others properties?) configurable (maybe at the IG level?). I could see a use case (say ephemeral pods for build agents on something like Jenkins) where an operator might want to not force the drain, but rather retry until ephemeral pods terminate naturally.

Reply:
Yes, I wanted to expose only Timeout, but can add Force as well.
I will also add some basic implementation of readiness gates

```yaml
spec:
  strategy:
    type: rollingUpdate
    rollingUpdate:
      maxUnavailable: 2
      readinessGates:
      # All pods in the namespace must be ready in order to proceed
      - type: Namespace
        value: namespace-1
      drainOptions:
        timeoutSeconds: 900
        force: true
```

One thing that is a bit challenging with this implementation is how to properly implement timeout for the batch of rotating nodes - since we requeue instead of blocking the thread while waiting for drain - might need to persist a timestamp in status, but that doesnt feel like the best solution.